source/mir/combinatorics/package.d

```diff
@@ -0,0 +1,184 @@
+module mir.combinatorics;
+
+import std.range.primitives: isRandomAccessRange, hasLength;
+import std.stdio;
+
+Permutations permutations(size_t n) @safe pure nothrow
+{
+    assert(n); //???
+    return Permutations(new size_t[n-1], new size_t[n]);
+}
+
+auto permutations(Range)(Range r) @safe pure nothrow
+if (isRandomAccessRange!Range && hasLength!Range)
+{
+    import std.range : indexed;
+    import std.algorithm: map;
+    size_t n = r.length;
+    assert(n); //???
+    return Permutations(new size_t[n-1], new size_t[n]).map!(p => r.indexed(p));
```

Comment:
`permutations` is the only method that is partially copied from Phobos -
https://dlang.org/library/std/algorithm/iteration/permutations.html

So I guess we have to provide a Range interface.